File: +shim/getcalmatrix.m
function A = getcalmatrix(F, H, S)
% function A = getcalmatrix(F, H, S)
%
% Get 2nd order shim calibration matrix.
% Not that this function 'quietly' inserts the DC (constant B0 offset) terms in F and S
% prior to calculating A.
%
% Inputs:
%   F    [N 8]      (Hz) Phase-unwrapped fieldmaps obtained by turning on/off each of the 8 (3 linear + 5 2nd order) shims.
%                   N = number of voxels ('control points')
%   H    [N ...]    SH basis matrix, evaluated at the same spatial locations (control points) as F.
%                   Unlike F and S, H also contains the DC term (1st column in H). See getSHshims.m.
%   S    [8 8]      (Pairwise difference in) shim currents used to obtain F (hardware units)
%   
% Output:
%   A    [9 9]      calibration matrix (includes DC term), expressed in hardware units

N = size(F,1);

% Add DC offset term (column) to F and S
% We do it here so the user doesn't have to remember to do it.
F = [ones(N,1) F]; 
Sin = S;
S = zeros(9);
S(1,1) = 1;
S(2:end,2:end) = Sin;

% get A
A = inv(H'*H)*H'*F*inv(S);                    % [9 9]
%A = inv(C'*W*W*C)*C'*W*W*F*inv(S);             % [9 9]

return

File: julia/getcalmatrix.jl
using LinearAlgebra

"""
    A = getcalmatrix(F::Array, H::Array, s::Vector)

In:
* `F` [nvoxel nshim] Acquired field maps
* `H` [nvoxel nbasis] Spatial basis functions
* `s` [nshim] applied shim currents (e.g., pairwise differences) used to acquire `F`
(excluding center frequency)
Out:
* `A` [nbasis nshim+1] (first column corresponds to center frequency)
"""
function getcalmatrix(F::AbstractMatrix{<:Real}, H::AbstractMatrix{<:Real}, s::AbstractVector)

	nb = size(H,2)      # number of basis functions. Includes center frequency.
	ns = length(s)      # number of shim terms (excluding center frequency)

	A = zeros(nb, ns+1)

	# center frequency (identity by definition)
	A[1,1] = 1.0

	# fit each acquired fieldmap to basis
	for ii = 1:ns
		A[:,ii+1] = s[ii] * H \ F[:,ii]
	end

	# A = inv(H'*H)*H'*F*inv(S);   [nb nb] Includes cf offset term.	

	return A
end
